### PR TITLE
[Rest Server] Check killAllOnCompletedTaskNumber in configuration

### LINKED ISSUE #622: killAllOnCompletedTaskNumber set to larger than the number of containers in the job
the job will never exit in this case.

### PR DESCRIPTION
**Pull request by abuccts**
Check killAllOnCompletedTaskNumber in configuration.

Fixes #622

### CODE REVIEW

File: rest-server/src/routes/job.js [hunk 1]
```diff
@@ -31,14 +31,14 @@ router.route('/')
     .get(param.jobQuery, jobController.list)
 
     /** POST /api/v1/jobs - Update job */
-    .post(tokenConfig.check, param.validate(jobConfig.schema), jobController.init, jobController.update);
+    .post(tokenConfig.check, param.validate(jobConfig.schema), param.checkKillAllOnCompletedTaskNumber, jobController.init, jobController.update);
```
Comment (fanyangCS) [inline rest-server/src/routes/job.js]: >checkKillAllOnCompletedTaskNumber [](start = 69, length = 33)

why don't we put checkKillAllOnCompletedTaskNumber into param.validate?

Comment (abuccts) [reply rest-server/src/routes/job.js]: `param.validate` is a general function which checks a json using a schema, it doesn't limit to job configuration

Comment (fanyangCS) [reply rest-server/src/routes/job.js]: it doens't seem to scale by defining a new function to check just one value

---
In reply to: [191151374](https://github.com/Microsoft/pai/pull/625#discussion_r191151374) [](ancestors = 191151374)

File: rest-server/src/middlewares/job.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+
+// module dependencies
+const param = require('./parameter');
+const logger = require('../config/logger');
+const jobConfig = require('../config/job');
+
+
+const checkKillAllOnCompletedTaskNumber = (req, res, next) => {
```
Comment (fanyangCS) [inline rest-server/src/middlewares/job.js]: >checkKillAllOnCompletedTaskNumber [](start = 6, length = 33)

maybe you can move this to parameter.js?

File: rest-server/src/middlewares/job.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      message: errorMessage,
+    });
+  } else {
+    next();
+  }
+};
+
+const submission = [
+  param.validate(jobConfig.schema),
+  checkKillAllOnCompletedTaskNumber,
```
Comment (fanyangCS) [inline rest-server/src/middlewares/job.js]: >checkKillAllOnCompletedTaskNumber [](start = 2, length = 33)

param.CheckKillAllOnCompletedTaskNumber()?

File: rest-server/src/routes/job.js [hunk 4]
```diff
@@ -19,26 +19,25 @@
...
-    .post(tokenConfig.check, param.validate(jobConfig.schema), jobController.init, jobController.update);
+    .post(tokenConfig.check, jobParam.submission, jobController.init, jobController.update);
 
 router.route('/:jobName')
     /** GET /api/v1/jobs/:jobName - Get job status */
     .get(jobController.get)
 
     /** PUT /api/v1/jobs/:jobName - Update job */
-    .put(tokenConfig.check, param.validate(jobConfig.schema), jobController.update)
+    .put(tokenConfig.check, jobParam.submission, jobController.update)
```
Comment (fanyangCS) [inline rest-server/src/routes/job.js]: >put [](start = 5, length = 3)

we will obsolete this API,right? please add a comment on it.

**Pull request merged**